external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutRetryExponentialBackoff.java

```diff
@@ -54,59 +56,92 @@
 
         public RetrySchedule(KafkaSpoutMessageId msgId, long nextRetryTime) {
             this.msgId = msgId;
-            this.nextRetryTimeNanos = nextRetryTime;
+            this.nextRetryTimeMs = nextRetryTime;
             LOG.debug("Created {}", this);
         }
 
         public void setNextRetryTime() {
-            nextRetryTimeNanos = nextTime(msgId);
+            nextRetryTimeMs = nextTime(msgId);
             LOG.debug("Updated {}", this);
         }
 
-        public boolean retry(long currentTimeNanos) {
-            return nextRetryTimeNanos <= currentTimeNanos;
+        public boolean retry(long currentTimeMs) {
+            return nextRetryTimeMs <= currentTimeMs;
         }
 
         @Override
         public String toString() {
             return "RetrySchedule{" +
                     "msgId=" + msgId +
-                    ", nextRetryTime=" + nextRetryTimeNanos +
+                    ", nextRetryTimeMs=" + nextRetryTimeMs +
                     '}';
         }
 
+        @Override
+        public int hashCode() {
+            int hash = 5;
+            hash = 29 * hash + Objects.hashCode(this.msgId);
+            return hash;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj) {
+                return true;
+            }
+            if (obj == null) {
+                return false;
+            }
+            if (getClass() != obj.getClass()) {
+                return false;
+            }
+            final RetrySchedule other = (RetrySchedule) obj;
+            if (!Objects.equals(this.msgId, other.msgId)) {
+                return false;
+            }
+            return true;
+        }
+
         public KafkaSpoutMessageId msgId() {
             return msgId;
         }
 
-        public long nextRetryTimeNanos() {
-            return nextRetryTimeNanos;
+        public long nextRetryTimeMs() {
+            return nextRetryTimeMs;
         }
     }
 
     public static class TimeInterval implements Serializable {
-        private final long lengthNanos;
-        private final long length;
-        private final TimeUnit timeUnit;
+        private final long lengthMs;
 
         /**
          * @param length length of the time interval in the units specified by {@link TimeUnit}
-         * @param timeUnit unit used to specify a time interval on which to specify a time unit
+         * @param timeUnit unit used to specify a time interval on which to specify a time unit. Smallest supported unit is milliseconds
          */
         public TimeInterval(long length, TimeUnit timeUnit) {
-            this.length = length;
-            this.timeUnit = timeUnit;
-            this.lengthNanos = timeUnit.toNanos(length);
+            
+            if(timeUnit == TimeUnit.MICROSECONDS || timeUnit == TimeUnit.NANOSECONDS) {
+                throw new IllegalArgumentException("TimeInterval does not support time units smaller than milliseconds");
+            }
```

Comment:
The class doesn't support time units less than milliseconds. I'd prefer to warn about misconfiguration rather than just clamping to 0 or 1 millisecond.